File: code/pigeonSort.cpp
void pigeonholeSort(int arr[], int n) {
    int MIN = arr[0], MAX = arr[0];

    for (int i = 1; i < n; i++) {
        MIN = std::min(MIN, arr[i]);
        MAX = std::max(MAX, arr[i]);
    }

    int range = MAX - MIN + 1;
    std::vector<int> holes(range, 0);

    for (int i = 0; i < n; i++)
        ++holes[arr[i] - MIN];

    int index = 0; 
    for (int i = 0; i < range; i++) {
        while (holes[i]--) {
            arr[index] = i + MIN;
            ++index;
        }
    }
}

File: code/pigeonSort.py
def pigeonhole_sort(arr):
    MIN = min(arr)
    MAX = max(arr)
    range_size = MAX - MIN + 1

    holes = [0] * range_size

    for num in arr:
        holes[num - MIN] += 1

    index = 0
    for i in range(range_size):
        while holes[i] > 0:
            arr[index] = i + MIN
            index += 1
            holes[i] -= 1